pkg/apis/core/types.go

```diff
@@ -3211,6 +3211,13 @@ type ServiceSpec struct {
 	// of peer discovery.
 	// +optional
 	PublishNotReadyAddresses bool
+
+	// topologyKeys is a preference-order list of topology keys.  If backends exist for
+	// index [0], they will always be chosen; only if no backends exist for index [0] will backends for index [1] be considered.
+	// If this field is specified and all indices have no backends, the service has no backends, and connections will fail.  We say these requirements are hard.
+	// In order to express soft requirement, we may give a special node label key "" as it means "match all nodes".
```

Comment:
"a soft requirement"

Reply:
We're making up syntax here.  "" seems somewhat obtuse.  What if we made the soft requirement be `"*"` rather than `""` ?  Thoughts?

@johnbelamaric

Reply:
Add 1 sentence about what happens if this is empty, e.g. "If this is not specified or empty, no topoly contraints will be applied".

Reply:
I like "*" better than "". It's at least an explicit indicator of "match any". However, it opens the door to people thinking they can do something like "zone/us-east-*". Which we don't want to do (I don't think...). So I would stick with just "".

Reply:
I think we can document "*" as "a special value, not a regex or glob"

Reply:
I still think it's asking for a few issues to get raised. People pick up by seeing it in examples, then they make an assumption; I think that's more likely than someone reading the docs.